### ISSUE TITLE
GPIO Edit for mobile GPUs

### ISSUE THREAD

**Issue by Panovich** [envytools/envytools#171]
I can't read or change the assignments for GPIO pins. I need to switch the mode from PWM to VID0-VID5, but nvbios does not read them correctly. GTX 950 m n 16 p-gt-a2 chip. And will there be a program to edit these values? I don't know how, where and what to change) Wanted to insert table from another BIOS.
[gx950m.zip](https://github.com/envytools/envytools/files/2424726/gx950m.zip)
`GPIO table at 0x5239 version 4.1
GPIO 9: line 9 tag 0x49 [THERM_ALERT_OUT] IN NEG DEF 0 gpio: normal SPEC_IN 0x15 [THERMAL_ALERT]
GPIO 11: line 11 tag 0x81 [VID_PWM0] OUT DEF 0 param 1 gpio: normal SPEC_OUT 0x5d [PWM_1]
GPIO 12: line 12 tag 0x6f [HW_PWR_SLOWDOWN] IN NEG DEF 0 gpio: normal
GPIO 13: line 13 tag 0x7a [ATX_FORCE_LOW_PWR] OUT NEG DEF 0 gpio: normal`
Should be:
0 - GC6_FB_EN
5 3v3_MAIN_EN
6 FB_CLAMP_REQ#
8 VGA_THERM#_R
9 ALERT#
10 FBREF_SEL
11 PWM_VID
12 DGPU_PD#
13 NVDD_PSI
16 FRM_CLK
21 GPU_PEX_RST_HOLD#
[parsed.zip](https://github.com/envytools/envytools/files/2424745/parsed.zip)
Parsed bios

**Comment by Echelon9** [comment]
Hi - Not sure I fully understand your question. Could you perhaps rephrase if English isn't your native language?

If you're asking whether envytools provides an ability to *edit* or *modify* dumped vbios files, that functionality is not currently included.

This repository is primarily used by developers of the `nouveau` userspace and kernel space driver.

**Comment by Panovich** [comment]
> Hi - Not sure I fully understand your question. Could you perhaps rephrase if English isn't your native language?
> 
> If you're asking whether envytools provides an ability to _edit_ or _modify_ dumped vbios files, that functionality is not currently included.
> 
> This repository is primarily used by developers of the `nouveau` userspace and kernel space driver.

Yes, English is not my native language. I used an interpreter. I wanted to point out that some gpio modes are not defined. Many simply skips, on some writes ???, but it is on most mobile chips. What is GC6 anyway ? I want to disable it, on my Board this is not :) open-gpu-doc says that it is memory-self-refresh, I can not understand what it means. And at the end of the BIOS there is some kind of certificate on his wrongness constantly writes nvflash. Can the video card not work because of the wrong certificate? All necessary changes made, the checksum recalculated. Crashes constantly when trying to use discrete graphics, video scheduler internal error, aida64 does not see the memory. Most likely the video BIOS is not correct.
